lnst/Tests/PktGen.py

```diff
@@ -0,0 +1,224 @@
+import re
+import time
+import logging
+from subprocess import Popen
+from threading import Thread
+from typing import Iterator, Union
+
+from lnst.Common.Utils import kmod_loaded
+from lnst.Common.IpAddress import Ip4Address
+from lnst.Tests.BaseTestModule import BaseTestModule, TestModuleError
+from lnst.Common.Parameters import IntParam, IpParam, StrParam, ListParam, DeviceParam
+
+
+class PktGenResultsParser:
+    def __init__(self, devs: list[str], duration: int) -> None:
+        """
+        PktGen output is just a table with current stats of devices. Therefore,
+        each device has a separate thread that captures current status of device
+        each second for `duration`.
+        """
+        self._devs = devs
+        self._duration = duration
+
+        self._sampling_threads: list[
+            tuple[str, Thread]
+        ] = []  # [(device, thread object)]
+        self._threads_start = {}  # {device: start_timestamp}
+        self._raw_samples = {}
+
+    def start_sampling(self):
+        """
+        This is a separate method just to emphasize that pktgen
+        needs to be started immediately after the start of sampling.
+        """
+        self._setup_capturing()
+
+        for device, thread in self._sampling_threads:
+            thread.start()
+            self._threads_start[device] = time.time()
+
+    def _setup_capturing(self):
+        for device in self._devs:
+            thread = Thread(target=self._read_dev_samples, args=(device,))
+            self._sampling_threads.append(
+                (device, thread)
+            )  # start of sampling timestamp is used for duration calculation of 1st sample
+
+    def _read_dev_samples(self, device: str):
+        self._raw_samples[device] = []
+
+        for _ in range(0, self._duration):
+            time.sleep(1)
```

Comment:
wouldn't this make more sense to be AFTER the proc file read? that way each thread immediately initializes with a raw sample which is basically "all zeros" and also stores the initial timestamp effectively removing the need for the `self._threads_start` variable and therefore also removing the conditional in the `parse_dev_outputs`?

Reply:
Hmm, thanks for pointing me in the right direction. That's a better solution.

And I also realized that I was actually saving **end** timestamp of sample instead of **start** one that is required due to time alignment/warm-up trimming, fixed.

Thanks.